### PR TITLE
fix #1804 Add a Scheduler that caps number of threads and scales to 0

### LINKED ISSUE #1804: Limit the number of threads created by the elastic scheduler
The current implementation of ElasticScheduler does not limit the number of threads created by it. 
Since we recommend using it for blocking APIs, it may lead to a problem when there are too many blocking calls and the scheduler will create 1000s of threads.

To avoid that, we should consider limiting the total number of threads and queue the tasks if all threads are busy.

### Things to consider
- this will be a behaviour change and **potentially** may break existing code, because if the tasks are scheduled recursively (a task submitted to the elastic scheduler submits a few more tasks to it) and wait for them, and the thread pool size is 100, then 101 task will wait forever because all other threads will be waiting
- Thread pool's size is always a pain to predict / do right

### Alternative
Introduce a new `Schedulers.blocking()` that will be elastic but limit the amount of threads (should be clearly stated in the docs, so that the users understand the potential of having a deadlock as described in "things to consider").

It could be a nice alternative given the elastic nature of both, and even there is some old code that uses `elastic`, their usage will not dramatically affect the app because their threads will eventually be terminated.

### PR DESCRIPTION
**Pull request by simonbasle**
Needs more tests.

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java [hunk 1]
```diff
@@ -151,19 +168,41 @@ public static Scheduler fromExecutorService(ExecutorService executorService, Str
...
+		return cache(CACHED_CAPPED, CAPPED, CAPPED_SUPPLIER);
+	}
+
 	/**
 	 * {@link Scheduler} that hosts a fixed pool of single-threaded ExecutorService-based
 	 * workers and is suited for parallel work.
 	 *
 	 * @return default instance of a {@link Scheduler} that hosts a fixed pool of single-threaded
-	 * ExecutorService-based workers
+	 * ExecutorService-based workers and is suited for parallel work
```
Comment (bsideup) [inline reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java]: we have to be careful with "parallel work" tho. People tend to think parallel == multi-threaded == heavy computations, while we actually use it for IO and non-blocking stuff

File: reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+			return e.cached;
+		}
+
+		if (REMAINING.decrementAndGet(this) < 0) {
+			//cap reached
+			REMAINING.incrementAndGet(this);
+			if (shutdown) {
+				return SHUTDOWN;
+			}
+			throw Exceptions.failWithRejected("cannot directly schedule task, cap of " + this.cap + " reached");
```
Comment (bsideup) [inline reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java]: don't we queue the tasks if we're at capacity?

File: reactor-core/src/test/java/reactor/core/scheduler/SchedulersMetricsTest.java [hunk 3]
```diff
@@ -182,41 +183,53 @@ public void disablingMetricsRemovesSchedulerMeters() {
...
-								.isEqualTo(taskCount);
-					});
+			assertThat(counters.stream()
+			                   .mapToDouble(FunctionCounter::count)
+			                   .sum())
+					.isEqualTo(taskCount);
 		});
     }
 
+    @Parameters(method = "metricsSchedulers")
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/scheduler/SchedulersMetricsTest.java]:
```suggestion
	@Parameters(method = "metricsSchedulers")
```

File: reactor-core/src/test/java/reactor/core/scheduler/SchedulersMetricsTest.java [hunk 4]
```diff
@@ -234,21 +247,20 @@ public void shouldReportExecutionTimes() {
 		}
 		phaser.arriveAndAwaitAdvance();
 
-		Timer timer = simpleMeterRegistry
+		Collection<Timer> timers = simpleMeterRegistry
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/scheduler/SchedulersMetricsTest.java]: Why `Collection`? We filter by tag, isn't it unique enough?

Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/scheduler/SchedulersMetricsTest.java]: no, for a given SCHEDULER_ID there are several executors for the elastic and capped cases (parallel is historically initialized at 1 for that test)

Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/scheduler/SchedulersMetricsTest.java]: and if there are multiple matching meters, `.timer()` will return the first matching timer, which is not satisfactory

File: reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java (hunk 2 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java]: only for the cases where these tasks are submitted through a `Worker`

File: reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java (hunk 1 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java]: I just realigned the javadocs for `elastic` and `parallel` flavors in here, but maybe in a separate follow up we could find a better wording.

File: reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java (hunk 2 above)
Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java]: Is there a strong reason for that? I am just thinking about some code that does `subscribeOn` inside a task and will fail, something I definitely wouldn't expect

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java]: `subscribeOn` uses `Worker`

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java]: the reason is practicality of implementation: all worker-based usage can be maintained consistent by providing the `DeferredWorker` that will queue the tasks. Free-form direct scheduling is much less obvious: should it queue in a common queue, a dedicated `DeferredWorker` that is only used for that purpose? round-robin to existing `DeferredWorker`s (with the caveat that this potentially impacts unrelated scheduling)? Reject? I simply chose the later because it was the simplest most deterministic solution.

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java]: (just got that you were talking about `FluxSubscribeOnValue` @bsideup)

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/scheduler/CappedScheduler.java]: I guess it could work that I `createWorker` on each direct schedule operation. Since the implementation calls `dispose()`, cancellation should correctly be propagated to the whole `Worker` through the returned `Disposable`.

**Comment by vy** [comment]
At [bol.com](bol.com), we have been a devoted RxJava (and later on, Reactor) user for 5+ years. I have witnessed many occasions where the elastic pool sizes and/or unbounded task queues constituted the perfect medium to hide almost every backpressure-related problem and created a temporal illusion of success that dreadfully collapses for long running applications in production.

To address this issue, in my new beloved project [reactor-pubsub](https://github.com/vy/reactor-pubsub) -- which has been blessed by @bsideup, of which I am deeply thankful -- I ship [`BoundedScheduledThreadPoolExecutor`](https://github.com/vy/reactor-pubsub/blob/master/src/main/java/com/vlkan/pubsub/util/BoundedScheduledThreadPoolExecutor.java): a  `ScheduledThreadPoolExecutor` wrapper with a bounded task queue. (The used method is hinted by Doug Lea. For details and further discussion see [the relevant concurrency-interest thread](http://cs.oswego.edu/pipermail/concurrency-interest/2019-April/016861.html).)

In summary, I will really appreciate a default Reactor `Scheduler` with a fixed thread count and bounded task queue. You can't imagine how much curse (since it exposes backpressure problems) and blessing (since it exposes backpressure problems) you will get from your users.

**Comment by simonbasle** [comment]
@vy ok, with `CappedScheduler` we're achieving a capping of threads, what would be missing here is the bounded queue aspect, which I guess can be made configurable in a `newCapped` factory. The only issue is that in order to ensure tasks submitted by an operator are run on a consistent `Worker`, we need a separate queue per worker. So it would be a bound to the queue size of each Worker, NOT for the whole `Scheduler`. does that sound like a good enough tradeoff in your experience?

**Comment by vy** [comment]
@simonbasle, infinitely better than the current proposal which provides no facilities for the task queue size. :champagne: In a nutshell, I will be able to configure the number of threads and the per-thread task queue size, correct?

Would you also mind explaining why do *"we need a separate queue per worker"*, please?

**Comment by smaldini** [comment]
We're going to collect users feedback during 3.3 and we might transition elastic to this. I've grown worried about `Schedulers.elastic` as well and this gives a nice alternative. Follow up work will need to expose the pending queue metrics correctly as well.
